src/nma/nma.rs

```diff
@@ -0,0 +1,383 @@
+// NAME Mips Assembler
+
+use crate::args::Args;
+use crate::u5::U5;
+use crate::u6::U6;
+use std::fs::File;
+use std::fs::{self, copy};
+use std::io::Write;
+use std::str;
+
+const R_EXPECTED_ARGS: usize = 3;
+
+enum R_form {
+    NONE,
+    RD_RS_RT,
+    RS,
+    RD_RT_SHAMT,
+}
+
+pub struct R {
+    shamt: U5,
+    funct: U6,
+    form: R_form,
+}
+
+const I_EXPECTED_ARGS: usize = 2;
+
+enum I_form {
+    NONE,
+    RT_IMM,
+}
+
+pub struct I {
+    opcode: U6,
+    form: I_form,
+}
+
+pub fn r_operation(mnemonic: &str) -> Result<R, &'static str> {
+    match mnemonic {
+        "add" => Ok(R {
+            shamt: U5 { value: 0 },
+            funct: U6 { value: 0x20 },
+            form: R_form::RD_RS_RT,
+        }),
+        "sub" => Ok(R {
+            shamt: U5 { value: 0 },
+            funct: U6 { value: 0x22 },
+            form: R_form::RD_RS_RT,
+        }),
+        "sll" => Ok(R {
+            shamt: U5 { value: 0 },
+            funct: U6 { value: 0x00 },
+            form: R_form::RD_RT_SHAMT,
+        }),
+        "srl" => Ok(R {
+            shamt: U5 { value: 0 },
+            funct: U6 { value: 0x02 },
+            form: R_form::RD_RT_SHAMT,
+        }),
+        "xor" => Ok(R {
+            shamt: U5 { value: 0 },
+            funct: U6 { value: 0x26 },
+            form: R_form::RD_RS_RT,
+        }),
+        _ => Err("Failed to match R-instr mnemonic"),
+    }
+}
+
+pub fn i_operation(mnemonic: &str) -> Result<I, &'static str> {
+    match mnemonic {
+        "lui" => Ok(I {
+            opcode: U6 { value: 0xf },
+            form: I_form::RT_IMM,
+        }),
+        _ => Err("Failed to match I-instr mnemonic"),
+    }
+}
+
+pub fn read_file(file_path: &str) -> String {
+    fs::read_to_string(file_path).expect("Failed to read the file")
```

Comment:
``.expect()`` immediately kills the program if an error occurs, let's not do that for error-prone filesystem access

Reply:
What should the correct course of action be here? Should I propagate a `Result` back up to `main`? For the other one I can just default to the existing assembler, although in the future I imagine the config file will become important enough that we should expect it

Reply:
Graceful fatal termination is the correct course of action IMO, the problem is that ``.expect()`` is not graceful and kills the program looking like this:

```
thread 'main' panicked at 'expect error message: "some message"', libcore/result.rs:945:5
```

That's ugly to users even though you are able to provide your own error message.

Reply:
Resolved